performance_test/src/communication_abstractions/ros2_waitset_communicator.hpp

```diff
@@ -0,0 +1,77 @@
+// Copyright 2017 Apex.AI, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef COMMUNICATION_ABSTRACTIONS__ROS2_WAITSET_COMMUNICATOR_HPP_
+#define COMMUNICATION_ABSTRACTIONS__ROS2_WAITSET_COMMUNICATOR_HPP_
+
+#include <rclcpp/rclcpp.hpp>
+
+#include <memory>
+#include <atomic>
+
+#include "../experiment_configuration/topics.hpp"
+#include "../experiment_configuration/qos_abstraction.hpp"
+
+#include "ros2_communicator.hpp"
+#include "resource_manager.hpp"
+
+namespace performance_test
+{
+
+/// Communication plugin for ROS 2 using waitsets for the subscription side.
+template<class Topic>
+class ROS2WaitsetCommunicator : public ROS2Communicator<Topic>
+{
+public:
+  /// The data type to publish and subscribe to.
+  using DataType = typename ROS2Communicator<Topic>::DataType;
+
+  /// Constructor which takes a reference \param lock to the lock to use.
+  explicit ROS2WaitsetCommunicator(SpinLock & lock)
+  : ROS2Communicator<Topic>(lock),
+    m_polling_subscription(nullptr) {}
+
+  /// Reads received data from ROS 2 using waitsets
+  void update_subscription() override
+  {
+    auto ros2QOSAdapter = this->m_ROS2QOSAdapter;
+    if (!m_polling_subscription) {
+      m_polling_subscription = this->m_node->template create_polling_subscription<DataType>(
+        Topic::topic_name() + this->m_ec.sub_topic_postfix(), *ros2QOSAdapter);
+    }
+    this->lock();
+    try {
+      rclcpp::Waitset<> m_waitset {m_polling_subscription};
+      auto wait_ret = m_waitset.wait(std::chrono::milliseconds(100));
+      if (wait_ret.all()) {
+        auto received_sample = m_polling_subscription->take();
+        DataType tmp = received_sample.data();
```

Comment:
If I want to use a reference, I would have to use a const reference since `m_polling_subscription->take()` returns an object of type `LoanedMessageT` which is a typedef of `class LoanedMessage` which has the `data()` function returning a const ref of type `MessageT` . I cannot assign `tmp` on line 59 above as a const ref due to https://github.com/ApexAI/performance_test/pull/88#discussion_r329335925

Reply:
I think this is not fixed yet.